plugins/maven/src/main/java/org/jetbrains/idea/maven/project/MavenProjectResolver.java

```diff
@@ -97,6 +111,150 @@ public void resolve(@NotNull Project project,
+
+    for (MavenProject mavenProject : requiredMavenProjects) {
+      mavenProject.setConfigFileError(null);
+
+      for (final MavenImporter mavenImporter : MavenImporter.getSuitableImporters(mavenProject)) {
+        mavenImporter.customizeUserProperties(project, mavenProject, userProperties);
+      }
+    }
+
+    final boolean updateSnapshots = mavenProjectsManager.getForceUpdateSnapshots() || generalSettings.isAlwaysUpdateSnapshots();
+    embedder.customizeForResolve(myTree.getWorkspaceMap(), console, process, updateSnapshots, userProperties);
+
+    try {
+      doResolve(project, requiredMavenProjects, generalSettings, embedder, context, process);
+    } catch (final Throwable t) {
+      final MavenConfigParseException cause = findParseException(t);
+
+      if (cause != null) {
+        MavenLog.LOG.warn("Cannot parse maven config", cause);
+      } else {
+        throw t;
+      }
+    } finally {
+      embeddersManager.release(embedder);
+    }
+
+    MavenUtil.restartConfigHighlightning(project, requiredMavenProjects);
+  }
+
+  @Nullable
+  private static MavenProjectManifestData parseManifest(@NotNull final MavenProject mavenProject) {
+    final VirtualFile directoryFile = mavenProject.getDirectoryFile();
+    final VirtualFile manifestFile = directoryFile.findFileByRelativePath("META-INF/MANIFEST.MF");
+
+    if (manifestFile == null) {
+      return null;
+    }
+
+    final Set<String> requiredBundles = new HashSet<>(32);
+    final Set<String> importedPackages = new HashSet<>(32);
+    final Set<String> exportedPackages = new HashSet<>(32);
+
+    try {
+      final Manifest manifest = new Manifest(manifestFile.getInputStream());
+      final Attributes manifestAttributes = manifest.getMainAttributes();
+      final String requiredBundlesStr = manifestAttributes.getValue("Require-Bundle");
+
+      if (requiredBundlesStr != null) {
+        for (final String bundle : requiredBundlesStr.split(",")) {
+          // Simply extract the Bundle name, discarding version requirements.
+          // Tycho will tell us if we have the right version or not
+          requiredBundles.add(bundle.split(";")[0].trim());
+        }
+      }
+
+      final String importedPackagesStr = manifestAttributes.getValue("Import-Package");
+
+      if (importedPackagesStr != null) {
+        for (final String packageName : importedPackagesStr.split(",")) {
+          importedPackages.add(packageName.trim());
+        }
+      }
+
+      final String exportedPackagesStr = manifestAttributes.getValue("Export-Package");
+
+      if (exportedPackagesStr != null) {
+        for (final String packageName : exportedPackagesStr.split(",")) {
+          exportedPackages.add(packageName.trim());
+        }
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return new MavenProjectManifestData(requiredBundles, importedPackages, exportedPackages);
+  }
+
+  private static void collectRequiredMavenProjects(
```

Comment:
This method uses recursion. It's fine for simplicity but it should be improved.
If by accident you have a circular dependency in MANIFEST files, you end up with a stack overflow.